R/get_vpts.R

```diff
@@ -0,0 +1,212 @@
+                  radar = string_extract(.data$directory, "(?<=hdf5\\/)[a-z]{5}"),
+                  date = as.Date(
+                    string_extract(.data$directory,
+                                   "[0-9]{4}\\/[0-9]{2}\\/[0-9]{2}$")
+                    )
+    )
+
+  # Check if the requested radars are present in the coverage
+  if(!all(selected_radars %in% coverage$radar)) {
+    cli::cli_abort(
+      "Radar(s) not found in ALOFT coverage:
+      {selected_radars[!selected_radars %in% coverage$radar]}.",
+      class = "getRad_error_radar_not_found")
+  }
+
+  # Check if the requested date radar combination is present in the coverage
+  ## We need to round the interval because coverage only has daily resolution
+  rounded_interval <-
+    lubridate::interval(
+      lubridate::floor_date(lubridate::int_start(date_interval), "day"),
+      lubridate::ceiling_date(lubridate::int_end(date_interval), "day")
+    )
+
+  at_least_one_radar_date_combination_exists <-
+    dplyr::filter(coverage,
+                radar %in% selected_radars,
+                date %within% rounded_interval) |>
+    nrow() > 0
+
+  if(!at_least_one_radar_date_combination_exists) {
+    cli::cli_abort(
+      "No data found for the requested radar(s) and date(s).",
+      class = "getRad_error_date_not_found")
+  }
+
+  # Filter the coverage data to the selected radars and time interval and
+  # convert into paths on the aloft s3 storage
+  ## We need to use the rounded interval because coverage only has daily
+  ## resolution
+  s3_paths <- dplyr::filter(coverage,
+                radar %in% selected_radars,
+                date %within% rounded_interval,
+                source %in% selected_sources) |>
+    dplyr::pull(.data$directory) |>
+    # Replace hdf5 with daily to fetch vpts files instead of hdf5 files
+    string_replace("hdf5", "daily") |>
+    # Construct the filename using regex magic!
+    string_replace(
+      "(.*?)/(.....)/(\\d{4})/(\\d{2})/(\\d{2})$",
+      "\\1/\\2\\/\\3/\\2_vpts_\\3\\4\\5.csv"
+    )
+
+  # Read the vpts csv files
+  aloft_data_url <-"https://aloftdata.s3-eu-west-1.amazonaws.com"
+
+  ## this could also be done by passing the vector of urls to readr::read_csv()
+  ## or vroom::vroom(), but both would be slower because they are not parallel
+  ## and wouldn't declare our custom user agent
+
+  vpts_from_s3 <-
+    paste(aloft_data_url, s3_paths, sep = "/") |>
+    purrr::map(httr2::request) |>
+    # Identify ourselves in the request
+    purrr::map(req_user_agent_getrad) |>
+    # Set retry conditions
+    purrr::map(~ httr2::req_retry(
+      .x,
+      max_tries = 15, backoff = \(x) sqrt(x) * 2,
+      is_transient = \(resp) resp_status(resp) %in% c(429),
+      retry_on_failure = TRUE
+    )) |>
+    # Perform the requests in parallel
+    httr2::req_perform_parallel() |>
+    # Fetch the response bodies and parse it using readr
+    purrr::map(httr2::resp_body_string) |>
+    purrr::map(~readr::read_csv(
```

Comment:
Is there a reason to use `readr` over `vroom` it seems vroom is still a bit faster (I think readr is just a wrapper around vroom), especially if you specify all columns.  Its also very memory efficient:
``` r
f <- "https://aloftdata.s3-eu-west-1.amazonaws.com/baltrad/daily/deumd/2024/deumd_vpts_20240102.csv"
download.file(f, t <- tempfile())
d <- vroom::vroom(t)
#> Rows: 7200 Columns: 26
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr   (2): radar, source_file
#> dbl  (21): height, u, v, w, ff, dd, sd_vvp, eta, dens, dbz, dbz_all, n, n_db...
#> lgl   (2): gap, vcp
#> dttm  (1): datetime
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
s <- vroom::spec(d)
bench::mark(
  check = TRUE, 
  vroom::vroom(t, show_col_types = F, progress = F),
  vroom::vroom(t, col_types = s, show_col_types = F, progress = F),
  readr::read_csv(t, show_col_types = F, progress = F),
  readr::read_csv(t, col_types = s, show_col_types = F, progress = F)
)
#> Registered S3 methods overwritten by 'readr':
#>   method                    from 
#>   as.data.frame.spec_tbl_df vroom
#>   as_tibble.spec_tbl_df     vroom
#>   format.col_spec           vroom
#>   print.col_spec            vroom
#>   print.collector           vroom
#>   print.date_names          vroom
#>   print.locale              vroom
#>   str.col_spec              vroom
#> # A tibble: 4 × 6
#>   expression                            min  median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                        <bch:t> <bch:t>     <dbl> <bch:byt>    <dbl>
#> 1 vroom::vroom(t, show_col_types =…  7.67ms  8.19ms     121.    96.02KB     6.50
#> 2 vroom::vroom(t, col_types = s, s…  5.77ms  6.35ms     156.    70.81KB     6.42
#> 3 readr::read_csv(t, show_col_type… 23.69ms 24.24ms      40.2    3.12MB     4.46
#> 4 readr::read_csv(t, col_types = s… 17.05ms 17.74ms      55.5    1.38MB     2.14
```

Reply:
I'm fine with switching to vroom